consult-search.el

```diff
@@ -0,0 +1,92 @@
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; consult-search integrates command line search tools and presents the results
+;; using completing-read.  This was written to work with consult
+
+;; This only works with ag for now
+
+;;; Code:
+
+(require 'projectile)
+
+(defun consult-search--marked-string ()
+  "Return the marked input (anything highlighted in buffer), nil if nothing is marked."
+  (when (use-region-p)
+    (buffer-substring-no-properties (region-beginning) (region-end))))
+
+(defun consult-search--search-string ()
+  "Get the input for search."
+  (or (consult-search--marked-string)
+      (thing-at-point 'symbol t)))
+
+(defun consult-search--do-ag (s path)
+  "Call ag with search string S on PATH and return all results.  Return a list of results (strings)."
+  (with-temp-buffer
+    (let ((exit-code (process-file "ag" nil t nil "--nocolor" "--nogroup" "--column" s path)))
+      (when (= exit-code 0)
+        (split-string (buffer-string) "\n" t)))))
+
+(defun consult-search--show-search-completions (search-string)
+  "Call ag with SEARCH-STRING on projectile's project root."
+  (completing-read
+   "ag: "
+   (consult-search--do-ag search-string (projectile-project-root))
+   nil t search-string))
+
+(defun consult-search--get-or-read-search-string ()
+  "Either get search string from or ask user for search string.  For use in interactive."
+  (list (or (consult-search--search-string) (read-string "Search for: "))))
+
+(defun consult-search--parse-filename-line-number (string)
+  "Parse filename and line number from STRING, which is assumed to be from ag output.  Input format is expected to be <file path>:<line number>:<column number>:<line from file>.  Return list of <file path>, <line number as int>, <column number as int>."
+  (let ((parts (split-string string ":")))
+    (list (car parts) (string-to-number (cadr parts)) (string-to-number (caddr parts)))))
+
+(defun consult-search--open-user-choice (string)
+  "Open user choice STRING, which is assumed to be from ag output."
+  (let* ((parsed (consult-search--parse-filename-line-number string))
+         (filename (car parsed))
+         (line-number (cadr parsed))
+         (column-number (caddr parsed)))
+    (find-file filename)
+    ;; Go back to the start of the file, then go to line-number
+    (goto-char (point-min)) (beginning-of-line line-number)
+    ;; Go forward to the right column
+    (forward-char (1- column-number)))
+  (let ((parts (split-string string ":")))
+    (list (car parts) (string-to-number (cadr parts)) (string-to-number (caddr parts)))))
+
+(defun consult-ag (search-string)
+  "Search for the marked string, or symbol at point, or if neither of the above are available, a SEARCH-STRING from prompt, then presents the user with search results from ag, and then opens the user's choice in the buffer."
```

Comment:
It's probably best to format this docstring to have a full sentence on the first line and rest broken down to more sentences.

Reply:
@manuel-uberti Please add more of your review comments! I would very much appreciate that - in case you have time to help out here!

Reply:
@minad sure! I'll see what I can do. And of course if `rg` comes into place here I'll be glad to try out the code just as I did for `consult-flymake`.

Reply:
@manuel-uberti Thanks!